internal/model/command.go

```diff
@@ -355,16 +356,39 @@ func runWithVersion(cmd *cobra.Command, artifactArch, desiredVersion string, sho
 
 	// If the workflow succeeded, make the used version the default
 	if shouldPromote && !env.IsGithubAction() && !env.IsLocalDev() {
-		currentExecPath, err := os.Executable()
-		if err != nil {
-			log.From(cmd.Context()).Warnf("failed to promote version: %s", err.Error())
-			return nil
+		if err := promoteVersion(cmd.Context(), vLocation); err != nil {
+			return fmt.Errorf("failed to promote version: %w", err)
 		}
+	}
 
-		if err := os.Rename(vLocation, currentExecPath); err != nil {
-			log.From(cmd.Context()).Warnf("failed to promote version: %s", err.Error())
-			return nil
-		}
+	return nil
+}
+func promoteVersion(ctx context.Context, vLocation string) error {
+	mutex := concurrency.NewIPMutex()
+	err := mutex.TryLock(ctx, func(attempt int) {
```

Comment:
Not sure about this. Why don't we lock just once at the start of `runWithVersion()`, more call sites of the same lock is more risk of deadlocks.